src/coreclr/vm/profilingenumerators.cpp

```diff
@@ -552,9 +552,11 @@ HRESULT ProfilerThreadEnum::Init()
     }
     CONTRACTL_END;
 
+    // If EnumThreads is called from a profiler callback where the runtime is already suspended,
+    // don't recursively acquire/release the ThreadStore Lock.
     // If a profiler has requested that the runtime suspend to do stack snapshots, it
     // will be holding the ThreadStore lock already
-    ThreadStoreLockHolder tsLock(!g_profControlBlock.fProfilerRequestedRuntimeSuspend);
+    ThreadStoreLockHolder tsLock(!GCHeapUtilities::IsGCInProgress() && !g_profControlBlock.fProfilerRequestedRuntimeSuspend);
```

Comment:
```suggestion
    ThreadStoreLockHolder tsLock(!ThreadStore::HoldingThreadStore() && !g_profControlBlock.fProfilerRequestedRuntimeSuspend);
```

Reply:
I've also looked at uses of `fProfilerRequestedRuntimeSuspend` and I think there are couple races around setting and clearing this flag.

The flag should be set _after_ the profiler succeeded at suspending the runtime, not before. 
https://github.com/dotnet/runtime/blob/016d356e43659b7789d23defe570628c8864ee73/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L6838-L6839

And the flag should be cleared _before_ resuming the runtime, not after. 
Otherwise if resuming thread is preempted right after it resumes the runtime, the flag could stay set for a very long time, while the threadstore is no longer owned by the profiler.
https://github.com/dotnet/runtime/blob/016d356e43659b7789d23defe570628c8864ee73/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L6877-L6878

Perhaps fix these races in the same change?

Reply:
I can swap the order in which `fProfilerRequestedRuntimeSuspend` is set. I don't think we have anything testing its behavior very well. Wouldn't setting the flag after `SuspendEE` returns also mean that any callback during `SuspendEE` wouldn't see that it was a profiler requested suspend? (which would be fine for this scenario as it would see that the ThreadStoreLock was already acquired)

Reply:
> Wouldn't setting the flag after SuspendEE returns also mean that any callback during SuspendEE wouldn't see that it was a profiler requested suspend?

I think until the suspend succeeded, it is not useful for other threads to know that profiler is trying to do it. At least not as a way to tell if it is safe to iterate the thread list. 
For the thread that is performing the suspension checking if it holds the lock is a more direct way to tell when it is ok to iterate. 

Perhaps we have other reasons why the order does not matter, but setting the flag before holding the lock seems at best confusing.

Reply:
It shouldn't matter to change the order. The only impact I can think of is that a profiler might get an error for a thread that has just been suspended, but a profiler should be calling SuspendRuntime and only starting sampling once that method has returned.

Reply:
Updated and swapped the order!